addons/html_editor/static/src/main/link/link_selection_plugin.js

```diff
@@ -41,105 +41,83 @@ export class LinkSelectionPlugin extends Plugin {
 
     handleCommand(command, payload) {
         switch (command) {
-            case "CLEAN":
-                this.clean(payload.root);
-                break;
             case "NORMALIZE":
                 this.normalize(payload.root || this.editable);
                 break;
+            case "CLEAN":
+                this.clean(payload.root);
+                break;
         }
     }
 
     /**
-     * Apply the o_link_in_selection class if the selection is in a single link,
-     * remove it otherwise.
-     *
-     * @param {EditorSelection} [selection]
+     * @param {Element} root
      */
-    resetLinkInSelection(selection = this.shared.getEditableSelection()) {
-        this.clearLinkInSelectionClass(this.editable);
-
-        const { anchorNode, focusNode } = selection;
-        const [anchorLink, focusLink] = [anchorNode, focusNode].map((node) =>
-            closestElement(node, "a")
-        );
-        const singleLinkInSelection = anchorLink === focusLink && anchorLink;
-
-        if (singleLinkInSelection && this.isLinkEligibleForZwnbsp(singleLinkInSelection)) {
-            singleLinkInSelection.classList.add("o_link_in_selection");
-        }
+    normalize(root) {
+        this.updateFEFFs(root);
+        this.resetLinkInSelection();
     }
 
-    isLinkEligibleForZwnbsp(link) {
-        return (
-            link.isContentEditable &&
-            this.editable.contains(link) &&
-            !this.resources.blacklistLinkZwnbsp?.some((callback) => callback(link))
-        );
+    /**
+     * @param {Element} root
+     */
+    clean(root) {
+        this.removeFEFFs(root);
+        this.clearLinkInSelectionClass(root);
+        this.removeEmptyLinks(root);
     }
 
-    // @todo: unify this and ZWSPlugin's cleanZWS into a single utility
-    removeZWNBSPs(textNode, cursors) {
-        const indexesToRemove = [...textNode.textContent.matchAll(/\uFEFF/g)].map(
-            (match) => match.index
-        );
-        for (const index of indexesToRemove.toReversed()) {
-            textNode.deleteData(index, 1);
-        }
-        cursors.update((cursor) => {
-            if (cursor.node === textNode) {
-                cursor.offset -= indexesToRemove.filter((i) => i < cursor.offset).length;
-            }
-        });
-    }
+    /**
+     * @param {Element} root
+     */
+    updateFEFFs(root) {
+        this.removeFEFFs(root, { exclude: isLegitZwnbsp });
 
-    clearLinkInSelectionClass(root) {
-        // @todo: maybe the querySelectorAll calls should include the root.
-        for (const link of root.querySelectorAll(".o_link_in_selection")) {
-            removeClass(link, "o_link_in_selection");
+        for (const link of selectElements(root, "a")) {
+            this.padLinkWithZws(link);
```

Comment:
Yes... indeed it would be better to only make one selection change instead of several ones.
Updating the cursor still must be done case by case (so inside `padLinkWithZws`), but I could call `cursors.restore` only once.
The reason I did not do it is because some tests fail. Those are tests where no links are involved and selection normalization is not expected. I would need to adapt these tests, OR pass around a variable `needToRestore` or something to check if any link was padded or not, and than conditionally call `cursors.restore` (which I don't like, it complicates the code...).

So, the advantage of calling `cursors.restore` in the inner function is that it does not get called if not necessary (.i.e no links involved in the normalization).
The disadvantage is that it gets called multiple times if multiple links are involved in the normalization.

Reply:
Ahh true, I didn't think that other tests would fail, and making those checks for links will add a lot of complexity